### ISSUE TITLE
Recursive alias does not match non generic subclass of generic type

### ISSUE THREAD

**Issue by jeromegn** [crystal-lang/crystal#1473]
(Was: Adding a K => V to a Hash subclass is not possible with certain types)

In short: With a subclassed (and permissive of many types) Hash, it's not always possible to add certain types of values to it.

Demonstation: http://play.crystal-lang.org/#/r/fbe

I lack the words to explain it better so hopefully the linked demo is self-explanatory.

Is this a bug, a feature or a missing feature?

**Comment by pgkos** [comment]
Probably related:
```crystal
ary = [] of Array(String | Char) | Array(Int32 | Float64)
ary << ["abc", 'd']  # works
ary << ["abc", "def"]  # does not compile
```
but this works:
```crystal
ary = [] of Array(String | Char)
ary << ["abc", 'd']  # works
ary << ["abc", "def"]  # also works
```

**Comment by jeromegn** [comment]
Something is odd. I just realized I'm able to assign a key to a `Array(BSON::Document)` fine when I generate the document from the bson source.

Essentially, this [this](https://github.com/jeromegn/bson.cr/blob/master/examples/basic.cr#L13-L21) value  of the document will make an `Array(BSON::Document | String | Int32 | Bool | BSON::ObjectId)`. Later on, it'll be [inserted in the document](https://github.com/jeromegn/bson.cr/blob/master/src/bson/document.cr#L20).

I'm not sure that would suddenly work...

I created a new playground code example http://play.crystal-lang.org/#/r/fix which shows how it works if the array's type is specified beforehand.

**Comment by asterite** [comment]
Closed in favor of #2665
